### PR TITLE
network: tc: add several settings for HTB

### LINKED ISSUE #15213: network: tc: htb: not all qdisc options is supported
Currently we can't setup: mtu, burst, cburst and r2q.

If htb class is 100Mbit (or more) and r2q in default value, in dmesg you will get:

```shell
[Tue Mar 24 17:11:07 2020] HTB: quantum of class 10001 is big. Consider r2q change.
```

### PR DESCRIPTION
**Pull request by yuwata**
Closes #15213.

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -3208,7 +3216,30 @@
...
           <para>Specifies the priority of the class. In the round-robin process, classes with the lowest
-          priority field are tried for packets first. This setting is mandatory.</para>
+          priority field are tried for packets first.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>Quantum=</varname></term>
+        <listitem>
+          <para>Specifies how much bytes to serve from leaf at once.</para>
```
Comment (Pesa) [inline man/systemd.network.xml]: s/much/many/

File: man/systemd.network.xml [hunk 2]
```diff
@@ -3208,7 +3216,30 @@
...
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>MTUBytes=</varname></term>
+        <listitem>
+          <para>Specifies the maximum packet size we create. When suffixed with K, M, or G, the specified
+          size is parsed as Kilobits, Megabits, or Gigabits, respectively, to the base of 1024.</para>
```
Comment (Pesa) [inline man/systemd.network.xml]: The option has "Bytes" in its name but size suffixes are interpreted as *bits*? very confusing

Comment (yuwata) [reply man/systemd.network.xml]: Ouch.

File: man/systemd.network.xml (hunk 1 above)
Comment (poettering) [reply man/systemd.network.xml]: the unit is bytes? hence should be called QuantumBytes=, no?

File: man/systemd.network.xml [hunk 3]
```diff
@@ -3192,6 +3192,14 @@
           to the class. Defaults to unset.</para>
         </listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>RateToQuantum=</varname></term>
+        <listitem>
+          <para>Takes an unsigned integer. The DRR quantums are calculated by <varname>Rate=</varname> /
```
Comment (poettering) [inline man/systemd.network.xml]: is that `/` supposed to mean division? if so, maybe spell this out as "divided by".

Maybe:

```
The DRR quantums are calculated by dividing the value configured in Rate= by RateToQuantum=.
```

or so

File: man/systemd.network.xml [hunk 4]
```diff
@@ -3208,7 +3216,30 @@
...
+          <para>Specifies the maximum packet size we create. When suffixed with K, M, or G, the specified
+          size is parsed as Kilobits, Megabits, or Gigabits, respectively, to the base of 1024.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>Overhead=</varname></term>
+        <listitem>
+          <para>Takes an unsigned integer which specifies per-packet size overhead used in rate
+          computations.</para>
```
Comment (poettering) [inline man/systemd.network.xml]: is this in bytes? if so, should be called OverheadBytes=

File: man/systemd.network.xml [hunk 5]
```diff
@@ -3230,6 +3261,24 @@
           is used.</para>
         </listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>Buffer=</varname></term>
```
Comment (poettering) [inline man/systemd.network.xml]: BufferBytes=

File: man/systemd.network.xml [hunk 6]
```diff
@@ -3230,6 +3261,24 @@
...
+        <listitem>
+          <para>Specifies the maximum bytes burst which can be accumulated during idle period. When suffixed
+          with K, M, or G, the specified size is parsed as Kilobits, Megabits, or Gigabits, respectively,
+          to the base of 1024.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>CeilBuffer=</varname></term>
```
Comment (poettering) [inline man/systemd.network.xml]: ditto

File: man/systemd.network.xml [hunk 7]
```diff
@@ -3230,6 +3261,24 @@
           is used.</para>
         </listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>Buffer=</varname></term>
+        <listitem>
+          <para>Specifies the maximum bytes burst which can be accumulated during idle period. When suffixed
+          with K, M, or G, the specified size is parsed as Kilobits, Megabits, or Gigabits, respectively,
+          to the base of 1024.</para>
```
Comment (poettering) [inline man/systemd.network.xml:3290]: first you say "bytes", then you go on with "kilobits"? is it bytes or bits?

File: src/network/tc/htb.c [hunk 8]
```diff
@@ -92,16 +95,80 @@ int config_parse_hierarchy_token_bucket_default_class(
...
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = qdisc_new_static(QDISC_KIND_HTB, network, filename, section_line, &qdisc);
+        if (r == -ENOMEM)
+                return log_oom();
```
Comment (ssahani) [inline src/network/tc/htb.c:122]: We should handle the `log_oom()` at ` qdisc = new(QDisc, 1);` as it's not a library. Here we should handle error `r < 0`. Or may be just OK with it.

Comment (yuwata) [reply src/network/tc/htb.c:122]: I will consider your point later in other PR. Thanks.

**Comment by yuwata** [comment]
@Pesa and @poettering I've force-pushed an updated version. I hope all points are addressed. PTAL.

**Pull request merged**